tests/integration/test_charm.py

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import json
+import logging
+import urllib.request
+from pathlib import Path
+
+import pytest
+import yaml
+from helpers import get_unit_address, update_status_freq  # type: ignore[attr-defined]
+
+log = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
+    """Build the charm-under-test and deploy it together with related charms.
+
+    Assert on the unit status before any relations/configurations take place.
+    """
+    # build and deploy charm from local source folder
+    charm_under_test = await ops_test.build_charm(".")
+    resources = {"alertmanager-image": "ubuntu/prometheus-alertmanager"}
+    await ops_test.model.deploy(charm_under_test, resources=resources, application_name="am")
```

Comment:
This is where `upstream-source` is really useful, as you don't have to specify the image being used again. I'm also bugging @johnsca to add some functionality to `pytest-operator` where you could write something like this:

```python
charm = await ops_test.build_charm()
await ops_test.model.deply(charm, resources=charm.default_resources())
```

Reply:
Also, since you're already parsing the `metadata.yaml` field as `METADATA`, you could just look up `upstream-source` from that instead of respecifying it here, as that tends to get forgotten about when updating `metadata.yaml`

Reply:
Thanks! Done.